tasks/AISTask.cpp

```diff
@@ -36,73 +36,147 @@ bool AISTask::configureHook()
+                    << e.what() << std::endl;
+        m_AIS_stats.invalid_messages++;
         return true;
     }
 
-    mAISStats.received_messages++;
+    m_AIS_stats.received_messages++;
     switch (msg->type()) {
         case ais::message_id::position_report_class_a: {
             auto msg01 = ais::message_cast<ais::message_01>(msg);
-            _positions.write(AIS::getPosition(*msg01));
+            auto position = AIS::getPosition(*msg01);
+            processPositionReport(position, position.mmsi);
             break;
         }
         case ais::message_id::position_report_class_a_assigned_schedule: {
             auto msg02 = ais::message_cast<ais::message_02>(msg);
-            _positions.write(AIS::getPosition(*msg02));
+            auto position = AIS::getPosition(*msg02);
+            processPositionReport(position, position.mmsi);
             break;
         }
         case ais::message_id::position_report_class_a_response_to_interrogation: {
             auto msg03 = ais::message_cast<ais::message_03>(msg);
-            _positions.write(AIS::getPosition(*msg03));
+            auto position = AIS::getPosition(*msg03);
+            processPositionReport(position, position.mmsi);
             break;
         }
         case ais::message_id::static_and_voyage_related_data: {
             auto msg05 = ais::message_cast<ais::message_05>(msg);
-            _vessels_information.write(AIS::getVesselInformation(*msg05));
-            _voyages_information.write(AIS::getVoyageInformation(*msg05));
+            auto vesselInformation = AIS::getVesselInformation(*msg05);
+            auto voyageInformation = AIS::getVoyageInformation(*msg05);
+            addToMap(vesselInformation);
+            _vessels_information.write(vesselInformation);
+            _voyages_information.write(voyageInformation);
             break;
         }
         default:
-            mAISStats.ignored_messages++;
+            m_AIS_stats.ignored_messages++;
             break;
     }
     return true;
 }
+void AISTask::processPositionReport(ais_base::Position& position, int mmsi)
+{
+    auto vessel = getCorrespondingVesselInfo(mmsi);
+    if (vessel.has_value()) {
+        Eigen::Quaterniond vessel2world_ori(
+            Eigen::AngleAxisd(position.course_over_ground.getRad(),
+                Eigen::Vector3d::UnitZ()));
+        auto sensor2world_pos =
+            sensorDataToWorld(vessel->reference_position, vessel2world_ori);
+        auto pos =
+            convertUTMToGPSInWorldFrame(convertGPSToUTM(position), sensor2world_pos);
+        position.latitude = pos.latitude;
+        position.longitude = pos.longitude;
+    }
+    _positions.write(position);
+}
+std::optional<ais_base::VesselInformation> AISTask::getCorrespondingVesselInfo(int mmsi)
+{
+    if (m_vessels.find(mmsi) != m_vessels.end()) {
+        return m_vessels.at(mmsi);
+    }
+
+    return std::nullopt;
+}
+void AISTask::addToMap(ais_base::VesselInformation info)
```

Comment:
```suggestion
void AISTask::updateKnownVessels(ais_base::VesselInformation info)
```

This way the purpose of this method is clear